source/common/http/utility.cc

```diff
@@ -761,6 +761,23 @@ const std::string& Utility::getProtocolString(const Protocol protocol) {
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
+std::string Utility::buildOriginalUri(const Http::RequestHeaderMap& request_headers,
+                                      const uint32_t max_path_length) {
+  if (!request_headers.Path()) {
+    return "";
+  }
+  absl::string_view path(request_headers.EnvoyOriginalPath()
+                             ? request_headers.getEnvoyOriginalPathValue()
+                             : request_headers.getPathValue());
+
+  if (path.length() > max_path_length) {
+    path = path.substr(0, max_path_length);
```

Comment:
What happens in practice if we truncate the path? Will consumers handle this gracefully? Maybe we should increment some stat at least to indicate that this is happening?

Reply:
So specifically, this code is based on how the `http_tracing` filter gets the original URL back from the request headers:
https://github.com/envoyproxy/envoy/blob/99ff096c1f032e8b83c21514f3d366a9e87526ea/source/common/tracing/http_tracer_impl.cc#L39

I'm fine with simplifying it if we understand the reduction in functionality we get by removing the extra checks. However, since this is in the common library and while this filter uses it in a way that the getEnvoyOriginalPath() check is not necessary, it may be necessary in the general case.

Truncating the path would return an "incomplete" URI. There's no defined length limit for URIs in the HTTP standard, though in practice most browser will balk at anything over 2048 characters. Consumers could set an arbitrarily high max_path_length or perhaps a flag could be added to say, I do not want truncation, if it matters to consumers. I do agree it would be nice to surface the fact that the truncation happened,

Reply:
If we're copying the implementation from another place to put it into common code, can we have that other location also use this shared code? The jwt filter will never see the original host header at the time this is called due to the call to buildUrl being in decodeHeaders, but if this is being used by other things I agree we should be incorporating the original host header (and ideally document this behavior as part of the docstring for this function). 

How about we make the trunaction flag optional (absl::optional<uint32_t>) and don't do truncation in this case unless required? Then the tracer lib can call this and preserve the trunacation behavior there while this new usage doesn't truncate until we know that this is necessary

Reply:
Oh and I would love to see unit tests testing this helper directly instead of via extensions, e.g. something in test/common/http/utility_test

Reply:
I agree on all points! I'll be working on this today